File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第5章寒假作业.java
//寒假作业 2016年第七届省赛C/C++A组  lanqiaoOJ 1388

public class Main {
    static int[] vis = new int[13];//用来表示是否用过
    static int[] b = new int[13];//用来存放数字
    static int ans = 0;
     
    public static void main(String[] args) {
        dfs(0);
        System.out.println(ans);
    }
     
    static void dfs(int step) {
        if (step>=3 && b[0]+b[1]!=b[2]) return;
        if (step>=6 && b[3]-b[4]!=b[5]) return;
        if (step>=9 && b[6]*b[7]!=b[8]) return;
        if (step>=12 && b[9]==b[10]*b[11]){
            ans++;
            return;
        }
         
        for (int i = 0; i < 13; i++) {
            if (vis[i] == 0) {
                b[step] = i + 1;
                vis[i] = 1;
                dfs(step + 1);//递归运行下一步
                vis[i] = 0;
            }
        }
    }
}

File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第5章寒假作业.py
#寒假作业 2016年第七届省赛C/C++A组  lanqiaoOJ 1388


ans=0
b=[0 for _ in range(15)]
vis=[0 for _ in range(15)]

def check3():  return b[1]+b[2] == b[3] 
def check6():  return b[4]-b[5] == b[6]
def check9():  return b[7]*b[8] == b[9] 
def check12(): return b[10]     == b[11]*b[12] 

def dfs(num):
    global ans
    if num==13: 
        if check12(): ans+=1
        return
    if num==4  and not check3():return
    if num==7  and not check6():return
    if num==10 and not check9():return
    for i in range(1,14):
        if not vis[i]:
            b[num]=i
            vis[i]=1
            dfs(num+1)
            vis[i]=0
              
dfs(1)
print(ans)
